web/main.js

```diff
@@ -109,40 +121,90 @@ let send = async f => {
 	sending = null;
 }
 
+let triggerDownload = receiving => {
+	if (serviceWorkerInUse) {
+		// `<a download=...>` doesn't work with service workers on Chrome yet.
+		// See https://bugs.chromium.org/p/chromium/issues/detail?id=468227
+
+		// Possible solutions:
+
+		// - `window.open` is blocked as a popup.
+		// window.open(`/download/${receiving.id}`);
+
+		// - And this is quite scary but `Content-Disposition` to the rescue!
+		//   It will navigate to 404 page if there is no service worker for some reason...
+		//   But if `postMessage` didn't throw we should be safe.
+		window.location = `/download/${receiving.id}`;
+
+	} else {
+		let blob = new Blob([receiving.data]);
+		let a = document.createElement('a');
+		a.href = URL.createObjectURL(blob); // TODO release this?
+		a.download = receiving.name;
+		a.style.display = 'none';
+		document.body.appendChild(a);
+		a.click();
+		document.body.removeChild(a);
+	}
+}
+
 // receive is the new message handler.
 //
 // This function cannot be async without carefully thinking through the
 // order of messages coming in.
 let receive = e => {
 	if (!receiving) {
 		receiving = JSON.parse(new TextDecoder('utf8').decode(e.data));
-		receiving.data = new Uint8Array(receiving.size);
+		receiving.id = document.getElementById("magiccode").value +
```

Comment:
sorry to come back to this. considering there is a (however unlikely and small) chance that this request ends up at the signalling server by accident, we shouldn not send it the magic code. the random number should be enough for now, we can use a guid in the future.

Reply:
(the wormhole code at this stage is already expired and useless, but still.)

Reply:
And those urls pollute browser history. Maybe file name + random string?

Reply:
sgtm